### DISCUSSION TITLE
issue printing ints from an array

### DISCUSSION THREAD

**Discussion by DrJaymz** [Bodmer/TFT_eSPI#2485] [Q&A - Sprites]
I've been using the library on a bunch of different hardware but I am having a silly issue with I'm 100% certain is caused by me.

I'm drawing a gauge into a sprite like this (don't need to go through that but note the for loop):

```
void drawGaugeSprite(M5Display &tft, int x, int y, int minValue, int maxValue, int value, bool isError, int labels[], int numLabels)
{
  int border = 3, pointerSize = 10;
  float pointer = (float)GAUGE_WIDTH / (float)maxValue;
  int top = GAUGE_HEIGHT / 2;
  int needle = (pointer * value);

  // draw outline
  sprite.fillSprite(BLACK);
  sprite.fillRect(0, 0, border, GAUGE_HEIGHT, WHITE);
  sprite.fillRect(GAUGE_WIDTH - border, 0, border, GAUGE_HEIGHT, WHITE);
  sprite.fillRect(0, GAUGE_HEIGHT - border, GAUGE_WIDTH, border, WHITE);

  // draw colour bar
  sprite.fillRect(border, top, GAUGE_WIDTH - border - border, (GAUGE_HEIGHT / 2) - border, GREEN);

  // draw labels
  sprite.textdatum = TC_DATUM;
  sprite.textsize = 1;
  for (int i = 0; i < numLabels; i++)
  {
    int pos = (int)(pointer * labels[i]);
    sprite.setCursor(pos, 5);

    char cstr[10];
    itoa(labels[i], cstr, 10);
    debug.printf("%i:%s ", labels[i], cstr);
    sprite.print(cstr);   //<-- crashes here with LoaderError
  }
  sprite.textdatum = TL_DATUM;

  // draw needle
  sprite.fillTriangle(needle - pointerSize, 0 - border, needle + pointerSize, 0 - border, needle, top, WHITE);

  // if data not valid then draw red cross over the gauge
  if (isError)
  {
    drawFatLineSprite(sprite, 0, 0, GAUGE_WIDTH, GAUGE_HEIGHT, 5, RED);
    drawFatLineSprite(sprite, 0, GAUGE_HEIGHT, GAUGE_WIDTH, 0, 5, RED);
  }

  // Copy the sprite to the screen
  sprite.pushSprite(x, y);
}
```

and calling it like this:
```
 int y = 20;

  int labels[] = {10, 30, 50, 70, 90, 100};
  drawGaugeSprite(M5.lcd, 100, y, 0, 120, (int)sensorData.fuelLitres, sensorData.fuelQtyError, labels, 6);
  M5.Lcd.setCursor(0, y);
  M5.Lcd.printf("Fuel Qty \n %i  ", (int)sensorData.fuelLitres);
```
You can see I'm using the M5 core2 but I see its subclassed the TFT library so I'm treating it as the same thing.

My intention is to iterate through the array of labels and when I get a match I print the label.
debug output would be:
10:10 30:30 50:50 70:70 90:90 100:100

This shows that I am not iterrating off the end and that I can unpack the labels I put in.

but both sprite.print(cstr) or sprite.print(labels[i]) cause it to crash. with 'loadprohibited'.

I thought maybe I had been getting confused with pointer and actual data which is why I added the debug to check that it was the data I was expecting and the debug shows what I expect.

I even used sprite.printf() so its exactly the same, but it still falls over.  So if printf works for serial port why does it crash for the sprite?

**Comment by Bodmer** [comment]
I don't provide support for M5 variants of the library. If you post a simple complete sketch that demonstrates the problem using this master library then you may get help.

I don't have access to any suitable hardware to test your sketch but maybe a library user can help.

**Reply by DrJaymz** [reply]
I have pulled that bit out and issue didn't occur.  So there's isn't anything fundamentally wrong with the way I'm calling it, or my syntax of the way I'm passing in the array?. 
The difference is that the array is now ints rather than short unsigned ints.  And I was creating and deleting the sprite and now I just keep reusing it.  So its a mystery.  If I still have issues, I can just pass in the increment (like every 20) and a starting figure.  But still would have liked to have understood why printf would work on serial and fail on sprite.

Thanks again for your as always swift response.

Can you tell me what the relationship is with M5, do they just maintain a fork of this library?  I only got a core 2 yesterday but soon was using it like any other board.

**Reply by Bodmer** [reply]
Great. It sounds like there is insufficient RAM and the memory is getting fragmented.

M5 copied and renamed the library. I don't know what version they use or what modifications have been made.
